### PR TITLE
feat(block): support prefix key encoding

### PR DESCRIPTION
**Pull request by KaguraMilet**
Solve #237.

Adding prefix key encoding for slateDB.

### CODE REVIEW

**Comment by KaguraMilet** [comment]
For the modification of `fn test_many_block_sst_iter` in `sst_iter.rs`, I found via dbg! that the block number was reduced to 5 due to the prefix key encoding.

Here are my results:
dbg

Test result:
test_result

File: src/sst_iter.rs [hunk 1]
```diff
@@ -329,7 +329,7 @@ mod tests {
             .unwrap();
         let sst_handle = table_store.open_sst(&SsTableId::Wal(0)).await.unwrap();
         let index = table_store.read_index(&sst_handle).await.unwrap();
-        assert_eq!(index.borrow().block_meta().len(), 6);
+        assert_eq!(index.borrow().block_meta().len(), 5);
```
Comment (criccomini) [inline src/sst_iter.rs:332]: Did this shrink because we're saving space on the key prefixes now?

**Review by criccomini — APPROVED** [review]
LGTM

File: src/sst_iter.rs (hunk 1 above)
Comment (KaguraMilet) [reply src/sst_iter.rs:332]: Based on my results using `dbg!`, yes. There is a screenshot of me above.

File: src/block.rs [hunk 2]
```diff
@@ -75,13 +93,15 @@ impl BlockBuilder {
     #[must_use]
     pub fn add(&mut self, key: &[u8], value: Option<&[u8]>) -> bool {
         assert!(!key.is_empty(), "key must not be empty");
+        let overlap_len = compute_prefix(&self.first_key, key);
```
[the remark this answers is not in the document]
Comment (KaguraMilet) [reply src/block.rs:96]: I need to look into their block_based_builder, but that needs to wait until my computer is fixed.

File: src/sst_iter.rs (hunk 1 above)
Comment (KaguraMilet) [reply src/sst_iter.rs:332]: Just add dbg! on L466 of `sst.rs`. You should be able to get the above results. I chose this line for convenience.

File: src/block.rs [hunk 3]
```diff
@@ -54,6 +54,22 @@ pub struct BlockBuilder {
     offsets: Vec<u16>,
     data: Vec<u8>,
     block_size: usize,
+    first_key: Bytes,
+}
+
+// Details can be found: https://users.rust-lang.org/t/how-to-find-common-prefix-of-two-byte-slices-effectively/25815/4
+fn compute_prefix(lhs: &[u8], rhs: &[u8]) -> usize {
```
Comment (rodesai) [inline src/block.rs:61]: it would be good to include a unit test for this

**Review by rodesai — COMMENTED** [review]
LGTM!

File: src/block_iterator.rs [hunk 4]
```diff
@@ -61,22 +64,32 @@ impl<B: BlockLike> KeyValueIterator for BlockIterator<B> {
...
+        let first_key = BlockIterator::decode_first_key(&block);
+
         let idx = block.offsets().partition_point(|offset| {
             let mut cursor = &block.data()[*offset as usize..];
-            let key_len = cursor.get_u16() as usize;
-            let cursor_key = &cursor[..key_len];
+            let overlap_len = cursor.get_u16() as usize;
+            let rest_len = cursor.get_u16() as usize;
+            let rest_key = &cursor[..rest_len];
+            let mut cursor_key = BytesMut::from(&first_key[..overlap_len]);
```
Comment (flaneur2020) [inline src/block_iterator.rs:84]: how about initialize `cursor_key` here with a `BytesMut::with_capacity`?

File: src/block_iterator.rs [hunk 5]
```diff
@@ -104,7 +118,22 @@ impl<B: BlockLike> BlockIterator<B> {
             ValueDeletable::Value(value)
         };
 
-        Some(KeyValueDeletable { key, value: v })
+        let mut key = BytesMut::from(&self.first_key[..overlap_len]);
+        key.extend_from_slice(&rest_key);
```
Comment (flaneur2020) [inline src/block_iterator.rs:122]: same as above, be like:

```
let mut key = BytesMut::with_capacity(overlap_len + rest_key.len());
key.extend_from_slice(&self.first_key[..overlap_len]);
key.extend_from_slice(&rest_key);
```

File: src/block_iterator.rs (hunk 4 above)
Comment (KaguraMilet) [reply src/block_iterator.rs:84]: Yes!

File: src/block_iterator.rs (hunk 5 above)
Comment (KaguraMilet) [reply src/block_iterator.rs:122]: I have to wait until my computer is fixed. I may modify the current implementation to perform delta encoding for `prev_key`.

**Comment by KaguraMilet** [comment]
I studied RocksDB's prefix key encoding. If we want to implement this function, we need to modify the `offsets` field of Block. For the new design, we use a new field called `restart_offsets`, each offset defines a new area which the key of the first entry in the area has no overlap with the previous area. This helps us quickly find `last_key` when rebuilding the complete key. Otherwise, it will be difficult to implement. So it depends on your idea whether to delta encode the first key of the block or the previous key of the inserted key.
@criccomini

**Comment by criccomini** [comment]
@KaguraMilet Let's stick with the current approach (first_key of block). We can modify things later if we find a need.

**Comment by criccomini** [comment]
@KaguraMilet looks like a test is failing on rebase. If you fix that and add a test for the `compute_prefix`, I think we're good to go.

**Comment by KaguraMilet** [comment]
> @KaguraMilet looks like a test is failing on rebase. If you fix that and add a test for the `compute_prefix`, I think we're good to go.

I think it's because the new test `block::tests::test_block_size` didn't take into account the key prefix encoding that caused the test to fail.
@criccomini

File: src/block_iterator.rs [hunk 6]
```diff
@@ -92,9 +105,10 @@
         let off_usz = off as usize;
         // TODO: bounds checks to avoid panics? (paulgb)
         let mut cursor = self.block.data().slice(off_usz..);
-        let key_len = cursor.get_u16() as usize;
-        let key = cursor.slice(..key_len);
-        cursor.advance(key_len);
+        let overlap_len = cursor.get_u16() as usize;
+        let rest_len = cursor.get_u16() as usize;
+        let rest_key = cursor.slice(..rest_len);
+        cursor.advance(rest_len);
```
[the remark this answers is not in the document]
Comment (criccomini) [reply src/block_iterator.rs:111]: Prefer an `assert` to the suggestion coderabbit made.

**Comment by KaguraMilet** [comment]
I fixed the problem, now it's time for me to go to bed. @criccomini

**Pull request merged**